File: src/org/kde/kdeconnect/Backends/BluetoothBackend/ConnectionMultiplexer.java
package org.kde.kdeconnect.Backends.BluetoothBackend;

import android.bluetooth.BluetoothSocket;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public final class ConnectionMultiplexer implements Closeable {
    private static final UUID DEFAULT_CHANNEL = UUID.fromString("a0d0aaf4-1072-4d81-aa35-902a954b1266");
    private static final int BUFFER_SIZE = 4096;

    private static final class ChannelInputStream extends InputStream implements Closeable {
        Channel channel;

        ChannelInputStream(Channel channel) {
            this.channel = channel;
        }

        @Override
        public int available() {
            return channel.available();
        }

        @Override
        public void close() throws IOException {
            channel.close();
        }

        @Override
        public int read() {
            byte[] b = new byte[1];
            if (read(b, 0, 1) == -1) {
                return -1;
            } else {
                return b[0];
            }
        }

        @Override
        public int read(byte[] b, int off, int len) {
            return channel.read(b, off, len);
        }

        @Override
        public int read(byte[] b) {
            return read(b, 0, b.length);
        }
    }

    private static final class ChannelOutputStream extends OutputStream implements Closeable {
        Channel channel;

        ChannelOutputStream(Channel channel) {
            this.channel = channel;
        }

        @Override
        public void close() throws IOException {
            channel.close();
        }

        @Override
        public void flush() throws IOException {
            channel.flush();
        }

        @Override
        public void write(byte[] b, int off, int len) throws IOException {
            channel.write(b, off, len);
        }

        @Override
        public void write(int b) throws IOException {
            byte[] data = new byte[1];
            data[0] = (byte) b;
            write(data, 0, 1);
        }

        @Override
        public void write(byte[] b) throws IOException {
            write(b, 0, b.length);
        }
    }

    private static final class Channel implements Closeable {
        ConnectionMultiplexer multiplexer;
        UUID id;
        ByteBuffer read_buffer = ByteBuffer.allocate(BUFFER_SIZE);
        final Object lock = new Object();
        boolean open = true;
        int requestedReadAmount = 0; //Number of times we requested some bytes from the channel
        int freeWriteAmount = 0; //Number of times we can safely send bytes over the channel

        Channel(ConnectionMultiplexer multiplexer, UUID id) {
            this.multiplexer = multiplexer;
            this.id = id;
        }

        int available() {
            synchronized (lock) {
                return read_buffer.position();
            }
        }

        public int read(byte[] b, int off, int len) {
            if (len == 0) return 0;

            while (true) {
                boolean makeRequest;
                synchronized (lock) {
                    if (read_buffer.position() >= len) {
                        read_buffer.flip();
                        read_buffer.get(b, off, len);
                        read_buffer.compact();

                        //TODO: non-blocking (opportunistic) read request
                        return len;
                    } else if (read_buffer.position() > 0) {
                        int numread = read_buffer.position();
                        read_buffer.flip();
                        read_buffer.get(b, off, numread);
                        read_buffer.compact();

                        //TODO: non-blocking (opportunistic) read request
                        return numread;
                    }

                    if (!open) return -1;
                    makeRequest = requestedReadAmount < BUFFER_SIZE;
                }

                if (makeRequest) {
                    multiplexer.readRequest(id);
                }

                synchronized (lock) {
                    if (!open) return -1;
                    if (read_buffer.position() > 0) continue;

                    try {
                        lock.wait();
                    } catch (Exception ignored) {}
                }
            }
        }

        @Override
        public void close() throws IOException {
            flush();
            synchronized (lock) {
                open = false;
                read_buffer.clear();
                lock.notifyAll();
            }
            multiplexer.closeChannel(id);
        }

        void doClose() {
            synchronized (lock) {
                open = false;
                lock.notifyAll();
            }
        }

        public void write(byte[] data, int off, int len) throws IOException {
            while (len > 0) {
                synchronized (lock) {
                    while (true) {
                        if (!open) throw new IOException("Connection closed!");

                        if (freeWriteAmount == 0) {
                            try {
                                lock.wait();
                            } catch (Exception ignored) {}
                        } else {
                            break;
                        }
                    }
                }

                int num_written = multiplexer.writeRequest(id, data, off, len);
                off += num_written;
                len -= num_written;
            }
        }

        void flush() throws IOException {
            multiplexer.flush();
        }
    }

    private BluetoothSocket socket;
    private Map<UUID, Channel> channels = new HashMap<>();
    private final Object lock = new Object();
    private boolean open = true;
    private boolean receivedProtocolVersion = false;

    private static final byte MESSAGE_PROTOCOL_VERSION = 0; //Negotiate the protocol version
    private static final byte MESSAGE_OPEN_CHANNEL = 1; //Open a new channel
    private static final byte MESSAGE_CLOSE_CHANNEL = 2; //Close a channel
    private static final byte MESSAGE_READ = 3; //Request some bytes from a channel
    private static final byte MESSAGE_WRITE = 4; //Write some bytes to a channel

    public ConnectionMultiplexer(BluetoothSocket socket) throws IOException {
        this.socket = socket;
        channels.put(DEFAULT_CHANNEL, new Channel(this, DEFAULT_CHANNEL));

        sendProtocolVersion();

        new ListenThread(socket).start();
    }

    private void sendProtocolVersion() throws IOException {
        byte[] data = new byte[23];
        ByteBuffer message = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN);

        message.put(MESSAGE_PROTOCOL_VERSION);
        message.putShort((short) 4);
        message.position(19);
        message.putShort((short) 1);
        message.putShort((short) 1);

        socket.getOutputStream().write(data);
    }

    private void handleException(IOException e) {
        synchronized (lock) {
            open = false;
            for (Channel channel : channels.values()) {
                channel.doClose();
            }
            channels.clear();
            if (socket.isConnected()) {
                try {
                    socket.close();
                } catch (IOException ignored) {
                }
            }
        }
    }

    private void closeChannel(UUID id) {
        synchronized (lock) {
            if (channels.containsKey(id)) {
                channels.remove(id);

                byte[] data = new byte[19];
                ByteBuffer message = ByteBuffer.wrap(data);
                message.order(ByteOrder.BIG_ENDIAN);
                message.put(MESSAGE_CLOSE_CHANNEL);
                message.putShort((short) 0);
                message.putLong(id.getMostSignificantBits());
                message.putLong(id.getLeastSignificantBits());

                try {
                    socket.getOutputStream().write(data);
                } catch (IOException e) {
                    handleException(e);
                }
            }
        }
    }

    private void readRequest(UUID id) {
        synchronized (lock) {
            Channel channel = channels.get(id);
            if (channel == null) return;

            byte[] data = new byte[21];
            synchronized (channel.lock) {
                if (!channel.open) return;
                if (channel.read_buffer.position() + channel.requestedReadAmount >= BUFFER_SIZE) return;
                int amount = BUFFER_SIZE - channel.read_buffer.position() - channel.requestedReadAmount;

                ByteBuffer message = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN);
                message.put(MESSAGE_READ);
                message.putShort((short) 2);
                message.putLong(id.getMostSignificantBits());
                message.putLong(id.getLeastSignificantBits());
                message.putShort((short) amount);
                channel.requestedReadAmount += amount;

                try {
                    socket.getOutputStream().write(data);
                } catch (IOException e) {
                    handleException(e);
                }
                channel.lock.notifyAll();
            }
        }
    }

    private int writeRequest(UUID id, byte[] write_data, int off, int write_len) throws IOException {
        synchronized (lock) {
            Channel channel = channels.get(id);
            if (channel == null) return 0;

            byte[] data = new byte[19 + BUFFER_SIZE];
            int length;
            synchronized (channel.lock) {
                if (!channel.open) return 0;
                if (channel.freeWriteAmount == 0) return 0;

                length = channel.freeWriteAmount;
                if (write_len < length) {
                    length = write_len;
                }

                ByteBuffer message = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN);
                message.put(MESSAGE_WRITE);
                //Convert length to signed short
                short lengthShort;
                if (length >= 0x10000) {
                    throw new IOException("Invalid buffer size, too large!");
                } else if (length >= 0x8000) {
                    lengthShort = (short) (-0x10000 + length);
                } else {
                    lengthShort = (short) length;
                }
                message.putShort(lengthShort);
                message.putLong(id.getMostSignificantBits());
                message.putLong(id.getLeastSignificantBits());

                message.put(write_data, off, length);

                channel.freeWriteAmount -= length;
                channel.lock.notifyAll();
            }

            try {
                socket.getOutputStream().write(data, 0, 19 + length);
            } catch (IOException e) {
                handleException(e);
            }
            return length;
        }
    }

    private void flush() throws IOException {
        synchronized (lock) {
            if (!open) return;
            socket.getOutputStream().flush();
        }
    }

    @Override
    public void close() throws IOException {
        if (socket == null) {
            return;
        }
        socket.close();
        socket = null;
        for (Channel channel : channels.values()) {
            channel.doClose();
        }
        channels.clear();
    }

    public UUID newChannel() throws IOException {
        UUID id = UUID.randomUUID();
        synchronized (lock) {
            byte[] data = new byte[19];
            ByteBuffer message = ByteBuffer.wrap(data);
            message.order(ByteOrder.BIG_ENDIAN);
            message.put(MESSAGE_OPEN_CHANNEL);
            message.putShort((short) 0);
            message.putLong(id.getMostSignificantBits());
            message.putLong(id.getLeastSignificantBits());

            try {
                socket.getOutputStream().write(data);
            } catch (IOException e) {
                handleException(e);
                throw e;
            }
            channels.put(id, new Channel(this, id));
        }
        return id;
    }

    public InputStream getDefaultInputStream() throws IOException {
        return getChannelInputStream(DEFAULT_CHANNEL);
    }

    public OutputStream getDefaultOutputStream() throws IOException {
        return getChannelOutputStream(DEFAULT_CHANNEL);
    }

    public InputStream getChannelInputStream(UUID id) throws IOException {
        synchronized (lock) {
            Channel channel = channels.get(id);
            if (channel == null) throw new IOException("Invalid channel!");
            return new ChannelInputStream(channel);
        }
    }

    public OutputStream getChannelOutputStream(UUID id) throws IOException {
        synchronized (lock) {
            Channel channel = channels.get(id);
            if (channel == null) throw new IOException("Invalid channel!");
            return new ChannelOutputStream(channel);
        }
    }

    private final class ListenThread extends Thread {
        InputStream input;
        OutputStream output;

        ListenThread(BluetoothSocket socket) throws IOException {
            input = socket.getInputStream();
            output = socket.getOutputStream();
        }

        private void read_buffer(byte[] buffer, int len) throws IOException {
            int num_read = 0;
            while (num_read < len) {
                int count = input.read(buffer, num_read, len - num_read);
                if (count == -1) {
                    throw new IOException("Couldn't read enough bytes!");
                }

                num_read += count;
            }
        }

        private void read_message() throws IOException {
            byte[] data = new byte[BUFFER_SIZE];
            read_buffer(data, 19);
            ByteBuffer message = ByteBuffer.wrap(data, 0, 19).order(ByteOrder.BIG_ENDIAN);
            byte type = message.get();
            int length = message.getShort();
            //signed short -> unsigned short (as int) conversion
            if (length < 0) length += 0x10000;
            long channel_id_msb = message.getLong();
            long channel_id_lsb = message.getLong();
            UUID channel_id = new UUID(channel_id_msb, channel_id_lsb);

            if (!receivedProtocolVersion && type != MESSAGE_PROTOCOL_VERSION) {
                throw new IOException("Did not receive protocol version message!");
            }

            if (type == MESSAGE_OPEN_CHANNEL) {
                synchronized (lock) {
                    channels.put(channel_id, new Channel(ConnectionMultiplexer.this, channel_id));
                }
            } else if (type == MESSAGE_CLOSE_CHANNEL) {
                synchronized (lock) {
                    Channel channel = channels.get(channel_id);
                    if (channel == null) return;
                    channels.remove(channel_id);

                    channel.doClose();
                }
            } else if (type == MESSAGE_READ) {
                if (length != 2) {
                    throw new IOException("Message length is invalid for 'MESSAGE_READ'!");
                }
                read_buffer(data, 2);
                int amount = ByteBuffer.wrap(data, 0, 2).order(ByteOrder.BIG_ENDIAN).getShort();
                //signed short -> unsigned short (as int) conversion
                if (amount < 0) amount += 0x10000;

                synchronized (lock) {
                    Channel channel = channels.get(channel_id);
                    if (channel == null) return;

                    synchronized (channel.lock) {
                        channel.freeWriteAmount += amount;
                        channel.lock.notifyAll();
                    }
                }
            } else if (type == MESSAGE_WRITE) {
                if (length > BUFFER_SIZE) {
                    throw new IOException("Message length is bigger than read size!");
                }
                read_buffer(data, length);

                synchronized (lock) {
                    Channel channel = channels.get(channel_id);
                    if (channel == null) return;
                    synchronized (channel.lock) {
                        if (channel.requestedReadAmount < length) {
                            throw new IOException("No outstanding read requests of this length!");
                        }
                        channel.requestedReadAmount -= length;
                        if (channel.read_buffer.position() + length > BUFFER_SIZE) {
                            throw new IOException("Shouldn't be getting more data when the buffer is too full!");
                        }
                        channel.read_buffer.put(data, 0, length);
                        channel.lock.notifyAll();
                    }
                }
            } else if (type == MESSAGE_PROTOCOL_VERSION) {
                //Allow more than 4 bytes data, for future extensibility
                if (length < 4) {
                    throw new IOException("Message length is invalid for 'MESSAGE_PROTOCOL_VERSION'!");
                }
                //We might need a larger buffer to read this
                if (length > data.length) {
                    data = new byte[1 << 16];
                }
                read_buffer(data, length);

                //Check remote endpoint protocol version
                int minimum_version = ByteBuffer.wrap(data, 0, 2).order(ByteOrder.BIG_ENDIAN).getShort();
                //signed short -> unsigned short (as int) conversion
                if (minimum_version < 0) minimum_version += 0x10000;
                int maximum_version = ByteBuffer.wrap(data, 2, 2).order(ByteOrder.BIG_ENDIAN).getShort();
                //signed short -> unsigned short (as int) conversion
                if (maximum_version < 0) maximum_version += 0x10000;

                if (minimum_version > 1 || maximum_version < 1) {
                    throw new IOException("Unsupported protocol version " + minimum_version + " - " + maximum_version + "!");
                }
                //We now support receiving other messages
                receivedProtocolVersion = true;
            } else {
                throw new IOException("Invalid message type " + (int) type);
            }
        }

        @Override
        public void run() {
            while (true) {
                synchronized (lock) {
                    if (!open) return;
                }
                try {
                    read_message();
                } catch (IOException e) {
                    handleException(e);
                    return;
                }
            }
        }
    }
}

File: src/main/java/org/kde/kdeconnect/backends/bluetooth/ConnectionMultiplexer.kt
/*
 * SPDX-FileCopyrightText: 2019 Matthijs Tijink <matthijstijink@gmail.com>
 * SPDX-FileCopyrightText: 2024 Rob Emery <git@mintsoft.net>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/
package org.kde.kdeconnect.backends.bluetooth

import android.bluetooth.BluetoothSocket
import android.util.Log
import org.kde.kdeconnect.helpers.ThreadHelper.execute
import java.io.Closeable
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class ConnectionMultiplexer(socket: BluetoothSocket) : Closeable {
    private class ChannelInputStream(val channel: Channel) : InputStream(), Closeable {
        override fun available(): Int {
            return channel.available()
        }

        @Throws(IOException::class)
        override fun close() {
            channel.close()
        }

        override fun read(): Int {
            val b = ByteArray(1)
            return if (read(b, 0, 1) == -1) {
                -1
            } else {
                b[0].toInt()
            }
        }

        override fun read(b: ByteArray, off: Int, len: Int): Int {
            return channel.read(b, off, len)
        }

        override fun read(b: ByteArray): Int {
            return read(b, 0, b.size)
        }
    }

    private class ChannelOutputStream(val channel: Channel) : OutputStream(), Closeable {
        @Throws(IOException::class)
        override fun close() {
            channel.close()
        }

        @Throws(IOException::class)
        override fun flush() {
            channel.flush()
        }

        @Throws(IOException::class)
        override fun write(b: ByteArray, off: Int, len: Int) {
            channel.write(b, off, len)
        }

        @Throws(IOException::class)
        override fun write(b: Int) {
            val data = ByteArray(1)
            data[0] = b.toByte()
            write(data, 0, 1)
        }

        @Throws(IOException::class)
        override fun write(b: ByteArray) {
            write(b, 0, b.size)
        }
    }

    private class Channel(val multiplexer: ConnectionMultiplexer, val id: UUID) : Closeable {
        val readBuffer: ByteBuffer = ByteBuffer.allocate(BUFFER_SIZE)
        val lock = ReentrantLock()
        var lockCondition: Condition = lock.newCondition()

        var open = true
        var requestedReadAmount = 0 //Number of times we requested some bytes from the channel
        var freeWriteAmount = 0 //Number of times we can safely send bytes over the channel
        fun available(): Int {
            lock.withLock { return readBuffer.position() }
        }

        fun read(b: ByteArray, off: Int, len: Int): Int {
            if (len == 0) return 0
            while (true) {
                var makeRequest: Boolean
                lock.withLock {
                    if (readBuffer.position() >= len) {
                        readBuffer.flip()
                        readBuffer[b, off, len]
                        readBuffer.compact()

                        //TODO: non-blocking (opportunistic) read request
                        return len
                    } else if (readBuffer.position() > 0) {
                        val numberRead = readBuffer.position()
                        readBuffer.flip()
                        readBuffer[b, off, numberRead]
                        readBuffer.compact()

                        //TODO: non-blocking (opportunistic) read request
                        return numberRead
                    }
                    if (!open) return -1
                    makeRequest = requestedReadAmount < BUFFER_SIZE
                }
                if (makeRequest) {
                    multiplexer.readRequest(id)
                }
                lock.withLock {
                    if (!open) return -1
                    if (readBuffer.position() <= 0) {
                        try {
                            lockCondition.await()
                        } catch (ignored: Exception) {
                        }
                    }
                }
            }
        }

        @Throws(IOException::class)
        override fun close() {
            flush()
            lock.withLock {
                if (!open) return
                open = false
                readBuffer.clear()
                lockCondition.signalAll()
            }
            multiplexer.closeChannel(id)
        }

        fun doClose() {
            lock.withLock {
                open = false
                lockCondition.signalAll()
            }
        }

        @Throws(IOException::class)
        fun write(data: ByteArray, off: Int, len: Int) {
            var offset = off
            var length = len
            while (length > 0) {
                lock.withLock {
                    while (true) {
                        if (!open) throw IOException("Connection closed!")
                        if (freeWriteAmount == 0) {
                            try {
                                lockCondition.await()
                            } catch (_: Exception) {
                            }
                        } else {
                            break
                        }
                    }
                }
                val numWritten = multiplexer.writeRequest(id, data, offset, length)
                offset += numWritten
                length -= numWritten
            }
        }

        @Throws(IOException::class)
        fun flush() {
            multiplexer.flush()
        }
    }

    private val socket: BluetoothSocket
    private val channels: MutableMap<UUID, Channel> = HashMap()
    private val channelsLock = ReentrantLock()
    private var open = true
    private var receivedProtocolVersion = false

    init {
        this.socket = socket
        channels[DEFAULT_CHANNEL] = Channel(this, DEFAULT_CHANNEL)
        sendProtocolVersion()
        execute(ListenRunnable(socket))
    }

    @Throws(IOException::class)
    private fun sendProtocolVersion() {
        val data = ByteArray(23)
        val message = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
        message.put(MESSAGE_PROTOCOL_VERSION)
        message.putShort(4.toShort())
        message.position(19)
        message.putShort(1.toShort())
        message.putShort(1.toShort())
        socket.outputStream.write(data)
    }

    private fun handleException(@Suppress("UNUSED_PARAMETER") ignored: Exception) {
        channelsLock.withLock {
            open = false
            for (channel in channels.values) {
                channel.doClose()
            }
            channels.clear()
            if (socket.isConnected) {
                try {
                    socket.close()
                } catch (_: IOException) {
                }
            }
        }
    }

    private fun closeChannel(id: UUID) {
        channelsLock.withLock {
            if (channels.containsKey(id)) {
                channels.remove(id)
                val data = ByteArray(19)
                val message = ByteBuffer.wrap(data)
                message.order(ByteOrder.BIG_ENDIAN)
                message.put(MESSAGE_CLOSE_CHANNEL)
                message.putShort(0.toShort())
                message.putLong(id.mostSignificantBits)
                message.putLong(id.leastSignificantBits)
                try {
                    socket.outputStream.write(data)
                } catch (e: IOException) {
                    handleException(e)
                }
            }
        }
    }

    private fun readRequest(id: UUID) {
        channelsLock.withLock {
            val channel = channels[id] ?: return
            val data = ByteArray(21)
            channel.lock.withLock {
                if (!channel.open) return
                if (channel.readBuffer.position() + channel.requestedReadAmount >= BUFFER_SIZE) return
                val amount = BUFFER_SIZE - channel.readBuffer.position() - channel.requestedReadAmount
                val message = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
                message.put(MESSAGE_READ)
                message.putShort(2.toShort())
                message.putLong(id.mostSignificantBits)
                message.putLong(id.leastSignificantBits)
                message.putShort(amount.toShort())
                channel.requestedReadAmount += amount
                try {
                    socket.outputStream.write(data)
                } catch (e: IOException) {
                    handleException(e)
                } catch (e: NullPointerException) {
                    handleException(e)
                }
                channel.lockCondition.signalAll()
            }
        }
    }

    @Throws(IOException::class)
    private fun writeRequest(id: UUID, writeData: ByteArray, off: Int, writeLen: Int): Int {
        channelsLock.withLock {
            val channel = channels[id] ?: return 0
            val data = ByteArray(19 + BUFFER_SIZE)
            var length: Int
            channel.lock.withLock {
                if (!channel.open) return 0
                if (channel.freeWriteAmount == 0) return 0
                length = channel.freeWriteAmount
                if (writeLen < length) {
                    length = writeLen
                }
                val message = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
                message.put(MESSAGE_WRITE)
                //Convert length to signed short
                val lengthShort: Short = if (length >= 0x10000) {
                    throw IOException("Invalid buffer size, too large!")
                } else if (length >= 0x8000) {
                    (-0x10000 + length).toShort()
                } else {
                    length.toShort()
                }
                message.putShort(lengthShort)
                message.putLong(id.mostSignificantBits)
                message.putLong(id.leastSignificantBits)
                message.put(writeData, off, length)
                channel.freeWriteAmount -= length
                channel.lockCondition.signalAll()
            }
            try {
                socket.outputStream.write(data, 0, 19 + length)
            } catch (e: IOException) {
                handleException(e)
            }
            return length
        }
    }

    @Throws(IOException::class)
    private fun flush() {
        channelsLock.withLock {
            if (!open) return
            socket.outputStream.flush()
        }
    }

    @Throws(IOException::class)
    override fun close() {
        channelsLock.withLock {
            socket.close()
            for (channel in channels.values) {
                channel.doClose()
            }
            channels.clear()
        }
    }

    @Throws(IOException::class)
    fun newChannel(): UUID {
        val id = UUID.randomUUID()
        channelsLock.withLock {
            val data = ByteArray(19)
            val message = ByteBuffer.wrap(data)
            message.order(ByteOrder.BIG_ENDIAN)
            message.put(MESSAGE_OPEN_CHANNEL)
            message.putShort(0.toShort())
            message.putLong(id.mostSignificantBits)
            message.putLong(id.leastSignificantBits)
            try {
                socket.outputStream.write(data)
            } catch (e: IOException) {
                handleException(e)
                throw e
            }
            channels.put(id, Channel(this, id))
        }
        return id
    }

    @get:Throws(IOException::class)
    val defaultInputStream: InputStream
        get() = getChannelInputStream(DEFAULT_CHANNEL)

    @get:Throws(IOException::class)
    val defaultOutputStream: OutputStream
        get() = getChannelOutputStream(DEFAULT_CHANNEL)

    @Throws(IOException::class)
    fun getChannelInputStream(id: UUID): InputStream {
        channelsLock.withLock {
            val channel = channels[id] ?: throw IOException("Invalid channel!")
            return ChannelInputStream(channel)
        }
    }

    @Throws(IOException::class)
    fun getChannelOutputStream(id: UUID): OutputStream {
        channelsLock.withLock {
            val channel = channels[id] ?: throw IOException("Invalid channel!")
            return ChannelOutputStream(channel)
        }
    }

    private inner class ListenRunnable(socket: BluetoothSocket) : Runnable {
        var input: InputStream = socket.inputStream
        var output: OutputStream = socket.outputStream

        @Throws(IOException::class)
        private fun readBuffer(buffer: ByteArray, len: Int) {
            var numRead = 0
            while (numRead < len) {
                val count = input.read(buffer, numRead, len - numRead)
                if (count == -1) {
                    throw IOException("Couldn't read enough bytes!")
                }
                numRead += count
            }
        }

        fun byteArrayToHexString(bytes: ByteArray): String {
            val sb = StringBuilder()
            for (b in bytes) {
                sb.append(String.format("0x%02x ", b.toInt() and 0xff))
            }
            return sb.toString()
        }

        @Throws(IOException::class)
        private fun readMessage() {
            var data = ByteArray(BUFFER_SIZE)
            readBuffer(data, 19)
            val message = ByteBuffer.wrap(data, 0, 19).order(ByteOrder.BIG_ENDIAN)
            val type = message.get()
            var length = message.short.toInt()
            //signed short -> unsigned short (as int) conversion
            if (length < 0) length += 0x10000
            val channelIdMostSigBits = message.long
            val channelIdLeastSigBits = message.long
            val channelId = UUID(channelIdMostSigBits, channelIdLeastSigBits)
            if (!receivedProtocolVersion && type != MESSAGE_PROTOCOL_VERSION) {
                Log.w("ConnectionMultiplexer", "Received invalid message '$message'")
                Log.w("ConnectionMultiplexer", "'data_buffer:(" + byteArrayToHexString(data) + ") ")
                Log.w("ConnectionMultiplexer", "as string: '$data' ")

                throw IOException("Did not receive protocol version message!")
            }
            when (type) {
                MESSAGE_OPEN_CHANNEL -> {
                    channelsLock.withLock {
                        channels.put(channelId, Channel(this@ConnectionMultiplexer, channelId))
                    }
                }
                MESSAGE_CLOSE_CHANNEL -> {
                    channelsLock.withLock {
                        val channel = channels[channelId] ?: return
                        channels.remove(channelId)
                        channel.doClose()
                    }
                }
                MESSAGE_READ -> {
                    if (length != 2) {
                        throw IOException("Message length is invalid for 'MESSAGE_READ'!")
                    }
                    readBuffer(data, 2)
                    var amount = ByteBuffer.wrap(data, 0, 2).order(ByteOrder.BIG_ENDIAN).short.toInt()
                    //signed short -> unsigned short (as int) conversion
                    if (amount < 0) amount += 0x10000
                    channelsLock.withLock {
                        val channel = channels[channelId] ?: return
                        channel.lock.withLock {
                            channel.freeWriteAmount += amount
                            channel.lockCondition.signalAll()
                        }
                    }
                }
                MESSAGE_WRITE -> {
                    if (length > BUFFER_SIZE) {
                        throw IOException("Message length is bigger than read size!")
                    }
                    readBuffer(data, length)
                    channelsLock.withLock {
                        val channel = channels[channelId] ?: return
                        channel.lock.withLock {
                            if (channel.requestedReadAmount < length) {
                                throw IOException("No outstanding read requests of this length!")
                            }
                            channel.requestedReadAmount -= length
                            if (channel.readBuffer.position() + length > BUFFER_SIZE) {
                                throw IOException("Shouldn't be getting more data when the buffer is too full!")
                            }
                            channel.readBuffer.put(data, 0, length)
                            channel.lockCondition.signalAll()
                        }
                    }
                }
                MESSAGE_PROTOCOL_VERSION -> {
                    //Allow more than 4 bytes data, for future extensibility
                    if (length < 4) {
                        throw IOException("Message length is invalid for 'MESSAGE_PROTOCOL_VERSION'!")
                    }
                    //We might need a larger buffer to read this
                    if (length > data.size) {
                        data = ByteArray(1 shl 16)
                    }
                    readBuffer(data, length)

                    //Check remote endpoint protocol version
                    var minimumVersion = ByteBuffer.wrap(data, 0, 2).order(ByteOrder.BIG_ENDIAN).short.toInt()
                    //signed short -> unsigned short (as int) conversion
                    if (minimumVersion < 0) minimumVersion += 0x10000
                    var maximumVersion = ByteBuffer.wrap(data, 2, 2).order(ByteOrder.BIG_ENDIAN).short.toInt()
                    //signed short -> unsigned short (as int) conversion
                    if (maximumVersion < 0) maximumVersion += 0x10000
                    if (minimumVersion > 1 || maximumVersion < 1) {
                        throw IOException("Unsupported protocol version $minimumVersion - $maximumVersion!")
                    }
                    //We now support receiving other messages
                    receivedProtocolVersion = true
                }
                else -> {
                    throw IOException("Invalid message type " + type.toInt())
                }
            }
        }

        override fun run() {
            while (true) {
                channelsLock.withLock {
                    if (!open) {
                        Log.w("ConnectionMultiplexer", "connection not open, returning")
                        return
                    }
                }
                try {
                    readMessage()
                } catch (e: IOException) {
                    Log.w("ConnectionMultiplexer", "run caught IOException", e)
                    handleException(e)
                    return
                }
            }
        }
    }

    companion object {
        private val DEFAULT_CHANNEL = UUID.fromString("a0d0aaf4-1072-4d81-aa35-902a954b1266")
        private const val BUFFER_SIZE = 4096
        private const val MESSAGE_PROTOCOL_VERSION: Byte = 0 //Negotiate the protocol version
        private const val MESSAGE_OPEN_CHANNEL: Byte = 1 //Open a new channel
        private const val MESSAGE_CLOSE_CHANNEL: Byte = 2 //Close a channel
        private const val MESSAGE_READ: Byte = 3 //Request some bytes from a channel
        private const val MESSAGE_WRITE: Byte = 4 //Write some bytes to a channel
    }
}
